taskcluster/translations_taskgraph/actions/train.py

```diff
@@ -452,38 +483,77 @@ def train_action(parameters, graph_config, input, task_group_id, task_id):
 
     parameters = dict(parameters)
 
-    start_stage = input.pop("start-stage", None)
-    if start_stage:
-        if "previous_group_ids" not in input:
-            raise Exception(
-                "'previous_group_ids' is required to use 'start-stage' (otherwise we can't skip earlier tasks)"
-            )
+    previous_group_ids = input.pop("previous-group-ids", None)
+    if previous_group_ids:
+        # Resume the pipeline by reusing the completed tasks from previous task groups
+        import taskcluster
+
+        TC_MOZILLA = "https://firefox-ci-tc.services.mozilla.com"
+        options = {"rootUrl": ("%s" % TC_MOZILLA)}
+        queue = taskcluster.Queue(options=options)
+        tasks_to_add = {}
+        for group_id in previous_group_ids:
+            group = queue.listTaskGroup(group_id)
+            for task in group["tasks"]:
+                if task["status"]["state"] != "completed":
+                    continue
+
+                task_id = task["status"]["taskId"]
+                task_name = task["task"]["metadata"]["name"]
+                # Skip service tasks
+                if (
+                    task_name.startswith("Action")
+                    or task_name.startswith("Decision")
+                    or task_name.startswith("PR")
+                ):
+                    continue
+
+                # Tasks from the latter previous-group-ids groups override previously found tasks
+                tasks_to_add[task_name] = task_id
+
+        logger.info(f"Found top level existing tasks`: {json.dumps(tasks_to_add, indent=2)}")
+
+        if tasks_to_add:
+            # Add ancestors of all the top level completed tasks
+            for task_id, label in get_ancestors(list(tasks_to_add.values())).items():
+                # Skip service tasks
+                if (
+                    label.startswith("Action")
+                    or label.startswith("Decision")
+                    or label.startswith("PR")
+                ):
+                    continue
+                tasks_to_add[label] = task_id
+
+            logger.info(f"Added ancestor tasks`: {json.dumps(tasks_to_add, indent=2)}")
+
+            # Optionally rerun the tasks with the specified prefix and their descendants by removing them from the existing tasks
```

Comment:
I think I'm a bit unclear on why you need to fetch descendants. As far as I understand things, they will always be rerun if one of their upstreams is rerun. I suspect you could get rid of the descendant walking and this would behave exactly the same way.

Am I missing something here?

Reply:
I didn't think about it this way. My goal was to produce the correct `existing_tasks` that include only the tasks that should be reused. You're right, it makes sense that if one of the intermediate steps is missing, all its dependencies in the existing tasks will be ignored and rescheduled. Do you think it makes sense to rely on this mechanism and simplify the code by removing the descendant retrieval code, or is it still better to produce the correct existing tasks if we already have the code to do this?

Reply:
You can keep it as is as far as I'm concerned, but unless I'm missing something, there's no way that any tasks that are descendants of one of the prefixed tasks would ever be re-used?